packages/workflow/src/internals.ts

```diff
@@ -534,14 +542,116 @@ export class Activator implements ActivationHandler {
+      // Update?
+      this.rejectUpdate(updateId, ApplicationFailure.nonRetryable(`Update has no handler: ${name}`));
+      return;
+    }
+    const execute = composeInterceptors(this.interceptors.inbound, 'handleUpdate', this.updateNextHandler.bind(this));
+
+    const validate = composeInterceptors(
+      this.interceptors.inbound,
+      'validateUpdate',
+      this.validateUpdateNextHandler.bind(this)
+    );
+
+    const makeInput = (): UpdateInput => ({
+      updateId,
+      args: arrayFromPayloads(this.payloadConverter, activation.input),
+      name,
+      headers: headers ?? {},
+    });
+
+    // The implementation below is responsible for upholding, and constrained
+    // by, the following contract:
+    //
+    // 1. During validation, any error must fail the Update; during the Update
+    //    itself, Temporal errors fail the Update whereas other errors fail the
+    //    activation.
+    //
+    // 2. The handler must not see any mutations of the arguments made by the
+    //    validator.
+    //
+    // 3. Any error when decoding/deserializing input must be caught and result
+    //    in rejection of the Update before it is accepted, even if there is no
+    //    validator.
+    //
+    // 4. The initial synchronous portion of the (async) Update handler should
+    //    be executed after the (sync) validator completes such that there is
+    //    minimal opportunity for a different concurrent task to be scheduled
+    //    between them.
+    //
+    // 5. The stack trace view provided in the Temporal UI must not be polluted
+    //    by promises that do not derive from user code. This implies that
+    //    async/await syntax may not be used.
+    //
+    // Note that there is a deliberately unhandled promise rejection below.
+    // These are caught elsewhere and fail the corresponding activation.
+    let input: UpdateInput;
+    try {
+      if (runValidator) {
+        validate(makeInput());
+      }
+      input = makeInput();
+    } catch (error) {
+      this.rejectUpdate(updateId, error);
+      return;
+    }
+    this.acceptUpdate(updateId);
+    untrackPromise(
+      execute(input)
+        .then((result) => this.completeUpdate(updateId, result))
+        .catch((error) => {
+          if (error instanceof TemporalFailure) {
+            this.rejectUpdate(updateId, error);
+          } else {
+            throw error;
+          }
+        })
+    );
+  }
+
+  // Intentionally non-async function so this handler doesn't show up in the
+  // stack trace. The user's handler may be a normal function or an async
+  // function.
+  protected updateNextHandler({ name, args }: UpdateInput): Promise<unknown> {
+    const entry = this.updateHandlers.get(name);
+    if (!entry) {
+      return Promise.reject(new IllegalStateError(`No registered update handler for update: ${name}`));
+    }
+    const { handler } = entry;
+    try {
+      return Promise.resolve(handler(...args));
```

Comment:
This doesn't do much, you need to await the handler or you'll get unhandled rejections.

Reply:
Hm, I would think that the errors are handled on line 614-619 above. Unhandled rejections do deliberately occur there (as I documented in the comment) and are caught by the global unhandled rejection handler which fails the WFT.

Reply:
OIC, `Promise.resolve` will link to the `handler` promise so that's fine.